src/main/java/net/oauth/jsontoken/discovery/AsyncVerifierProviders.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package net.oauth.jsontoken.discovery;
+
+import com.google.common.collect.Maps;
+import net.oauth.jsontoken.AsyncJsonTokenParser;
+import net.oauth.jsontoken.crypto.SignatureAlgorithm;
+import net.oauth.jsontoken.crypto.Verifier;
+import java.util.Map;
+
+/**
+ * The asynchronous counterpart of {@link VerifierProviders}.
+ * A collection of {@link AsyncVerifierProvider}s, one for each signature algorithm.
+ * The {@link AsyncJsonTokenParser} uses a {@link AsyncVerifierProviders} instance to locate
+ * verification keys. In particular, it will first look up the {@link AsyncVerifierProvider}
+ * for the signature algorithm used in the JSON Token (different signature methods
+ * will use different ways to look up verification keys - for example, symmetric keys
+ * will always be pre-negotiated and looked up in a local database, while public
+ * verification keys can be looked up on demand), and the ask the {@link AsyncVerifierProvider}
+ * to provide a future that will return a {@link Verifier} to check the validity of the JSON Token.
+ */
+public class AsyncVerifierProviders {
```

Comment:
Since we're creating a new type anyway, we should avoid having a mutable class here. Either just make this an interface with a single method getVerifierProvider, or make it a concrete final class with a Builder that exposes addVerifierProvider(SignatureAlgorithm, AsyncVerifierProvider).

Reply:
I think having an interface here would provide more flexibility for whoever uses the library. They would be able to make it immutable or mutable, depending on their needs.